src/shared/misc.c

```diff
@@ -43,6 +43,94 @@
 
 extern char **environ;
 
+char *
+rc_deptree_cache_path_get(void) {
+	static char path[PATH_MAX] = { 0 };
```

Comment:
Returning mutable `static` arrays are just as bad as globals. Either let the caller supply the buffer (better, IMO) or return a malloc-ed pointer (worse, complicates memory management).

One other possibility in this case is having some sort of `init` function which sets these up, and then letting the user just (read-only) access those directly.

Reply:
The static was supposed to be immutable (aside from a set function), but I see how that can't really be enforced. A buffer works better indeed, the only annoyance would be having to re-create the string every time, for a thing that should be set basically only once.

Reply:
Something like this would work:

```c
struct dirs { const char *x_dir; const char *y_dir; ... };

struct dirs dirs_create(void); // allocates the strings inside the struct and returns it
void dirs_destroy(struct dirs *p); // `free`s all the stuff
```

And then the caller can call `dirs_create()` at startup/init and use that struct for the entire program and call `dirs_destroy()` before exiting.

Reply:
While that would work, there's a lot of disconnect places calling for rc_.*_get() functions, so those strings would still be re-generated quite a lot... If I'm not mistaken, I see three solutions:

* Just create a new one in many places anyway, then we could either use a struct like that, or have buffers as arguments
* Keep with static/globals, with the risks of callers modifying them by accident
* Rework quite a *lot* of functions to accept a struct dir* as a parameter

I don't quite like the third option, so would the overhead of regenerating the same strings over and over a few times, be worth the lessen chance of an accidental bug? (Or am I not seeing a forth option?)

Reply:
Ok, I rewrote the whole thing, and taking this into consideration!